### ISSUE TITLE
Variadic System Functions: Principles?

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#1325]
In the current spec, the following functions are variadic:

* `fn:concat`
* `fn:codepoints-to-string`
* `fn:distinct-ordered-nodes`

The advantage of the variadic representation is that a user can omit additional parentheses; the drawback is that the function cannot be enhanced with parameters later on. In my point of view, `fn:distinct-ordered-nodes` might be a candidate for that in the future (on the other hand, it will only be used by a very small user group one or the other way).

Other candidates for variadicity could be (among others):

* `fn:count`
* `fn:exists`, `fn:empty`
* `fn:head`, `fn:tail`, `fn:trunk`, `fn:foot`
* `fn:one-or-more`, `fn:exactly-one`, `fn:zero-or-one`
* `fn:innermost`, `fn:outermost`, `fn:unordered`, `fn:reverse`
* `fn:boolean`, `fn:not`
* `fn:identity`
* `fn:data`, `fn:has-children` (the 0-arity case would need to be preserved due to its special semantics)

Note that function calls like `fn:exactly-one(E1, E2)` can be reasonable in practice, as `E1` may return an empty sequence.

I wonder if we can define a principle which functions should be variadic?

**Comment by MarkNicholls** [comment]
@ChristianGruen 

personally I'm not (currently) a fan, I think the advantages are minimal at the cost of a nuanced overloading of syntax, I consciously try to ignore variadic functions.

I applaud things that make the language easier to use, just not variadic functions, maybe there's a killer user case I'm missing.

**Comment by michaelhkay** [comment]
I don't think we would be doing this if it weren't for `concat()`. It just feels so ugly to have a feature that only works for one particular function.

**Comment by MarkNicholls** [comment]
agreed but what if you think doing it for concat was a mistake though?

you can't undo it, but extending the principle would be throwing good money after bad.

For me the sequence operator `(,)` is there to unify an unbounded recursive definition into a single type and convenient syntax, variadic functions, seem to then let the let the unbounded type genie out of the bottle and reinvent the wheel in order to save a pair of parenthesis. Or am I missing the point?

**Comment by ChristianGruen** [comment]
I think that variadicity for `fn:identity` could be helpful. We already have many uses of `fn:identity#1 in the spec. `fn:identity#2` (`fn:identity#3`, etc.) could be supplied as a function argument to return more than 1 argument unchanged.

Apart from that, I share some of Mark’s concerns. It’s a frequently observed mistake in XPath that people forget the additional parens when passing sequences. However, if we offer some functions that support variadicity while others don’t, confusion might even increase.

Some thoughts on particular functions:

* `fn:concat` gets less relevant, now that we have various types of string constructors.
* `fn:string-join` could be another candidate, but it’s not possible to change it due the separator argument.
* `array:join` could be another one… unless we don’t enhance it with a separator (#868).
* `fn:data`… `fn:data#0` will always have a special meaning, as it processes the current context value.
* `fn:empty`… `fn:empty(A, B, C)` could also imply that `true` is returned if at least one argument is empty.
* `fn:codepoints-to-string`… doesn’t hurt, but we lose the possibility to enhance it with an options arguments later on (i.e., to decide how to handle invalid codepoints).
* Similarly, `fn:distinct-ordered-nodes`: With an options arguments, the processing of nodes in different documents could be controlled.

Thinking more about, I would tend to only offer variadicity for `fn:concat` and `fn:identity`.

**Comment by michaelhkay** [comment]
The functions I would most like to make variadic are those that are often applied to just two items as input: for example min(), max(), map:merge(). If my own experience is anything to go by, leaving out the brackets on these is a very common error. The trouble is that in all these cases there is an extra (and optional) options or collation argument that prevents it. There doesn't seem an easy answer to this one.

**Comment by MarkNicholls** [comment]
You may be able to get around the issues of optional options or multiple variadic parameters by currying your function signatures.

i.e. tupled

`foo((1,2,3),('a','b','c'))
`
vs curried

`foo(1,2,3)('a','b','c')
`
but you probably then need to invent syntax to say a parameter isn't variadic, I think you're robbing peter to pay paul.

I'm not suggesting it!

This conversation (and some others) remind me of 

"There are two ways of constructing a software design: One way is to make it so simple that there are obviously no deficiencies, and the other way is to make it so complicated that there are no obvious deficiencies. The first method is far more difficult"

Variadic signatures, to me, feel like the latter, it makes the model more complex to remove a class of deficiencies (in humans), I think the biggest gift you can give users who are relatively new to the language is the former, if they miss out the brackets, its an error (obviously).

**Comment by rhdunn** [comment]
I'd be happy with keeping this syntax to a limited set of functions where it makes most sense.

As noted, the motivation for this was `fn:concat` being written in the F&O spec as having a special `...` syntax at the end of the function that wasn't specified anywhere -- I don't like the idea of a language's function APIs using a feature that is not available in the core language. Other vendors (e.g. BaseX and MarkLogic) have provided variadic functions in their XQuery libraries. However, it was not possible to define your own variadic functions.

**Comment by ChristianGruen** [comment]
Thanks everyone for the feedback.

After further reflection, the best solution seems to me to restrict variadicity of built-in functions to `fn:concat`:

* Built-in functions should be extensible in the future (variadicity would spoil that).
* Arbitrary variadicity will increase confusion as long as we don’t have clear rules which functions are variadic and which are not.
* Historically, `fn:concat` needs to be varadic, so we cannot change it.

I believe that the situation is different for user-defined functions: In a defined scope, it can be helpful to make specific functions varadic.

**Comment by ndw** [comment]
At [meeting 092](https://qt4cg.org/meeting/minutes/2024/10-01.html), there was some sentiment in favor of leaving this issue open. Several members supported the notion that a list of functions that *should* be variadic should be drawn up and principles derived from this list.
